crates/chia-datalayer/src/merkle/blob.rs

```diff
@@ -0,0 +1,2539 @@
+            "expected leaf for index from key cache: {index} -> <<self>>"
+        ));
+
+        Ok((index, leaf, block))
+    }
+
+    pub fn get_parent_index(&self, index: TreeIndex) -> Result<Parent, Error> {
+        Ok(self.get_block(index)?.node.parent())
+    }
+
+    pub fn get_lineage_with_indexes(
+        &self,
+        index: TreeIndex,
+    ) -> Result<Vec<(TreeIndex, Node)>, Error> {
+        let mut next_index = Some(index);
+        let mut lineage = vec![];
+
+        while let Some(this_index) = next_index {
+            let node = self.get_node(this_index)?;
+            next_index = node.parent().0;
+            lineage.push((this_index, node));
+        }
+
+        Ok(lineage)
+    }
+
+    pub fn get_lineage_indexes(&self, index: TreeIndex) -> Result<Vec<TreeIndex>, Error> {
+        let mut next_index = Some(index);
+        let mut lineage: Vec<TreeIndex> = vec![];
+
+        while let Some(this_index) = next_index {
+            lineage.push(this_index);
+            next_index = self.get_parent_index(this_index)?.0;
+        }
+
+        Ok(lineage)
+    }
+
+    // pub fn iter(&self) -> MerkleBlobLeftChildFirstIterator<'_> {
+    //     <&Self as IntoIterator>::into_iter(self)
+    // }
+
+    pub fn calculate_lazy_hashes(&mut self) -> Result<(), Error> {
+        // OPT: yeah, storing the whole set of blocks via collect is not great
+        for item in LeftChildFirstIterator::new_with_block_predicate(
+            &self.blob,
+            None,
+            Some(|block: &Block| block.metadata.dirty),
+        )
+        .collect::<Vec<_>>()
+        {
+            let (index, mut block) = item?;
+            assert!(block.metadata.dirty);
+
+            let Node::Internal(ref leaf) = block.node else {
+                panic!("leaves should not be dirty")
+            };
+            // OPT: obviously inefficient to re-get/deserialize these blocks inside
+            //      an iteration that's already doing that
+            let left_hash = self.get_hash(leaf.left)?;
+            let right_hash = self.get_hash(leaf.right)?;
+            block.update_hash(&left_hash, &right_hash);
+            self.insert_entry_to_blob(index, &block)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn get_keys_values(&self) -> Result<HashMap<KeyId, ValueId>, Error> {
+        let mut map = HashMap::new();
+        for (key, index) in self.block_status_cache.iter_keys_indexes() {
+            let node = self.get_node(*index)?;
+            let leaf = node.expect_leaf(
+                "key was just retrieved from the key to index mapping, must be a leaf",
+            );
+            map.insert(*key, leaf.value);
+        }
+
+        Ok(map)
+    }
+
+    pub fn get_key_index(&self, key: KeyId) -> Result<TreeIndex, Error> {
+        self.block_status_cache
+            .get_index_by_key(key)
+            .copied()
+            .ok_or(Error::UnknownKey(key))
+    }
+
+    pub fn get_proof_of_inclusion(
```

Comment:
an idea for a fuzzer:
generate a random tree, ensure that every node can successfully generate a proof-of-inclusion (and maybe even validate the proof)

Reply:
- [ ] https://github.com/Chia-Network/chia_rs/pull/1205